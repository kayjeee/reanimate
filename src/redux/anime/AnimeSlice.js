import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import axios from 'axios';

const baseUrl = 'https://kitsu.io/api/edge';

const initialState = {
  animeList: [],
  anime: [],
  loading: true,
};

export const fetchAnime = createAsyncThunk('anime/fetchAnime', async (category) => {
  try {
    const response = await axios.get(`${baseUrl}/anime?filter[categories]=${category}&page[limit]=20`);
    const aList = response.data.data.map((anime) => ({
      id: anime.id,
      ...anime.attributes,
    }));
    return aList;
  } catch (error) {
    console.error('Failed to fetch the anime using id:', error);
    throw error;
  }
});

export const fetchAnimeById = createAsyncThunk('anime/fetchAnimeById', async (animeId) => {
  try {
    const response = await axios.get(`${baseUrl}/anime/${animeId}`);
    const aData = {
      id: response.data.data.id,
      ...response.data.data.attributes,
    };
    return aData;
  } catch (error) {
    console.error('Failed to fetch the anime using ID:', error);
    throw error;
  }
});

const animeSlice = createSlice({
  name: 'anime',
  initialState,
  reducers: {},
  extraReducers: (builder) => {
    builder
      .addCase(fetchAnime.pending, (state) => {
        state.loading = true;
      })
      .addCase(fetchAnime.fulfilled, (state, action) => {
        state.loading = false;
        state.animeList = action.payload;
      })
      .addCase(fetchAnime.rejected, (state) => {
        state.loading = false;
      })
      .addCase(fetchAnimeById.pending, (state) => {
        state.loadingAnime = true;
      })
      .addCase(fetchAnimeById.fulfilled, (state, action) => {
        state.loadingAnime = false;
        state.anime = action.payload;
      })
      .addCase(fetchAnimeById.rejected, (state) => {
        state.loadingAnime = false;
      });
  },
});

export default animeSlice.reducer;
