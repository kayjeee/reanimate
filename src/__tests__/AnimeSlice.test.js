/**
 * @jest-environment jsdom
 */
import axios from 'axios';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { fetchAnimeById } from '../redux/anime/AnimeSlice';

// Create a mock store with Redux middleware
const mockStore = configureMockStore([thunk]);
const store = mockStore();

describe('fetchAnimeById', () => {
  beforeEach(() => {
    // Clear the actions of the mock store before each test
    store.clearActions();
  });

  it('fetch anime and return anime', async () => {
    const animeId = '123';
    const expectedAnimeData = {
      id: animeId,
      title: 'Test Anime',
      description: 'anime test',
    };

    // Mock the axios.get function to return a resolved Promise with the expected anime data
    jest.spyOn(axios, 'get').mockResolvedValue({
      data: {
        data: {
          id: animeId,
          attributes: { ...expectedAnimeData },
        },
      },
    });

    // Dispatch the fetchAnimeById action
    await store.dispatch(fetchAnimeById(animeId));

    // Compare the actions dispatched by the store with the snapshot
    expect(store.getActions()).toMatchSnapshot();
  });

  it('should handle error if anime fetch by ID fails', async () => {
    const animeId = '123';

    // Mock the axios.get function to return a rejected Promise with an error
    jest.spyOn(axios, 'get').mockRejectedValue(new Error('Anime fetch failed'));

    // Dispatch the fetchAnimeById action
    await store.dispatch(fetchAnimeById(animeId));

    // Compare the actions dispatched by the store with the snapshot
    expect(store.getActions()).toMatchSnapshot();
  });
});
