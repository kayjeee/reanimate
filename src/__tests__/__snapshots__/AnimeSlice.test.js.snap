// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchAnimeById fetch anime and return anime 1`] = `
Array [
  Object {
    "meta": Object {
      "arg": "123",
      "requestId": "xcAGPjxpmJ9Wjtn3QGgOR",
      "requestStatus": "pending",
    },
    "payload": undefined,
    "type": "anime/fetchAnimeById/pending",
  },
  Object {
    "meta": Object {
      "arg": "123",
      "requestId": "xcAGPjxpmJ9Wjtn3QGgOR",
      "requestStatus": "fulfilled",
    },
    "payload": Object {
      "description": "anime test",
      "id": "123",
      "title": "Test Anime",
    },
    "type": "anime/fetchAnimeById/fulfilled",
  },
]
`;

exports[`fetchAnimeById should handle error if anime fetch by ID fails 1`] = `
Array [
  Object {
    "meta": Object {
      "arg": "123",
      "requestId": "jNycGExXs5xRDyna4cyr8",
      "requestStatus": "pending",
    },
    "payload": undefined,
    "type": "anime/fetchAnimeById/pending",
  },
  Object {
    "error": Object {
      "message": "Anime fetch failed",
      "name": "Error",
      "stack": "Error: Anime fetch failed
    at Object.<anonymous> (C:\\\\Users\\\\kagis\\\\reanimate\\\\src\\\\__tests__\\\\AnimeSlice.test.js:48:48)
    at Promise.then.completed (C:\\\\Users\\\\kagis\\\\reanimate\\\\node_modules\\\\jest-circus\\\\build\\\\utils.js:391:28)
    at new Promise (<anonymous>)
    at callAsyncCircusFn (C:\\\\Users\\\\kagis\\\\reanimate\\\\node_modules\\\\jest-circus\\\\build\\\\utils.js:316:10)
    at _callCircusTest (C:\\\\Users\\\\kagis\\\\reanimate\\\\node_modules\\\\jest-circus\\\\build\\\\run.js:218:40)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at _runTest (C:\\\\Users\\\\kagis\\\\reanimate\\\\node_modules\\\\jest-circus\\\\build\\\\run.js:155:3)
    at _runTestsForDescribeBlock (C:\\\\Users\\\\kagis\\\\reanimate\\\\node_modules\\\\jest-circus\\\\build\\\\run.js:66:9)
    at _runTestsForDescribeBlock (C:\\\\Users\\\\kagis\\\\reanimate\\\\node_modules\\\\jest-circus\\\\build\\\\run.js:60:9)
    at run (C:\\\\Users\\\\kagis\\\\reanimate\\\\node_modules\\\\jest-circus\\\\build\\\\run.js:25:3)
    at runAndTransformResultsToJestFormat (C:\\\\Users\\\\kagis\\\\reanimate\\\\node_modules\\\\jest-circus\\\\build\\\\legacy-code-todo-rewrite\\\\jestAdapterInit.js:170:21)
    at jestAdapter (C:\\\\Users\\\\kagis\\\\reanimate\\\\node_modules\\\\jest-circus\\\\build\\\\legacy-code-todo-rewrite\\\\jestAdapter.js:82:19)
    at runTestInternal (C:\\\\Users\\\\kagis\\\\reanimate\\\\node_modules\\\\jest-runner\\\\build\\\\runTest.js:389:16)
    at runTest (C:\\\\Users\\\\kagis\\\\reanimate\\\\node_modules\\\\jest-runner\\\\build\\\\runTest.js:475:34)
    at Object.worker (C:\\\\Users\\\\kagis\\\\reanimate\\\\node_modules\\\\jest-runner\\\\build\\\\testWorker.js:133:12)",
    },
    "meta": Object {
      "aborted": false,
      "arg": "123",
      "condition": false,
      "rejectedWithValue": false,
      "requestId": "jNycGExXs5xRDyna4cyr8",
      "requestStatus": "rejected",
    },
    "payload": undefined,
    "type": "anime/fetchAnimeById/rejected",
  },
]
`;
